### ISSUE TITLE
Bindings must be consts

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#1229]
Our syntax for creating a binding is currently:

```
bind(Foo).toClass(Bar);
```

This does not produce a constant value in Dart and wherefore not usable as an annotation value. In JS it is not statically analyzable by any tool.

The culprit is the `Binding` class. It will become a const class with all fields `@FINAL` and constructor `@CONST`. For one, we need to stop eagerly computing the `key` from `token` and constructing dependency list. Example:

```
export class Binding {
  @FINAL
  token;
  @FINAL
  factory:Function;
  @FINAL
  dependencies:List; // of types, not dependencies
  @FINAL
  providedAsPromise:boolean;

  @CONST
  constructor(token, factory:Function, dependencies:List, providedAsPromise:boolean) {
    this.token = token;
    this.factory = factory;
    this.dependencies = dependencies;
    this.providedAsPromise = providedAsPromise;
  }
}
```

The rest of the code needs to be adjusted accordingly.

**Comment by rkirov** [comment]
Can you add an example of how the const API would look like? We might need to keep the shorthand `bind(...).to(...)` in JS-land if the const dart API is too verbose.

**Comment by yjbanov** [comment]
I propose that this:

```
bind(Foo).toClass(Bar);
bind(Foo).toFactory(func, const [dep1, dep2]);
bind(Foo).toValue(someConstValue);
bind(Foo).toAlias(AnotherFoo);
```

Becomes this in Dart:

```dart
const Bind(Foo, toClass: Bar);
const Bind(Foo, toFactory: func, const [dep1, dep2]);
const Bind(Foo, toValue: someConstValue);
const Bind(Foo, toAlias: AnotherFoo);
```

And becomes this in JS:

```dart
new Bind(Foo, {toClass: Bar});
new Bind(Foo, {toFactory: func, [dep1, dep2]});
new Bind(Foo, {toValue: someConstValue});
new Bind(Foo, {toAlias: AnotherFoo});
```

Note that this is necessary for static analysis in both Dart and JS. We will loose a lot of toolability without it. I'm personally willing to pay a little in verbosity if it makes the syntax more toolable.

**Comment by mhevery** [comment]
I vote that we keep current in JS and use const in Dart:

JavaScript
```
bind(Foo).toClass(Bar);
bind(Foo).toFactory(func, const [dep1, dep2]);
bind(Foo).toValue(someConstValue);
bind(Foo).toAlias(AnotherFoo);
```

Dart:
```
const Bind(Foo, toClass: Bar);
const Bind(Foo, toFactory: func, deps: const [dep1, dep2]);
const Bind(Foo, toValue: someConstValue);
const Bind(Foo, toAlias: AnotherFoo);
```

**Comment by pkozlowski-opensource** [comment]
I'm with @mhevery here. Tooling is very important but having intuitive, human writable / readable syntax is even more important IMO.

**Comment by yjbanov** [comment]
Not sure what's not readable about the proposed syntax, but I'll defer the judgement about the JS-side to you. Regardless, the API for Dart still needs to be fixed. There's currently no way to define bindings in Dart's annotations.

**Comment by mhevery** [comment]
@yjbanov :+1: for dart fix

**Comment by yjbanov** [comment]
Fixed by #1286
